onnxruntime/core/session/inference_session.h

```diff
@@ -109,9 +110,12 @@ class InferenceSession {
     Create a new InferenceSession
     @param session_options Session options.
     @param session_env This represents the context for the session and contains the logger and the global threadpools.
+    @param prepacked_weights_container An optional container to store the prepacked weights of shared initializers 
+                                       to be shared across sessions.
     */
   explicit InferenceSession(const SessionOptions& session_options,
-                            const Environment& session_env);
+                            const Environment& session_env,
+                            PrepackedWeightsContainer* prepacked_weights_container = nullptr);
 
```

Comment:
As this isn't needed at construction time, I'd suggest something more like the setup for registering EPs. Added to OrtSessionOptions for external APIs, added to InferenceSession via a separate method before InferenceSession::Initialize is called.

Reply:
That makes ownership even more confusing in my opinion. To the external world, it looks like the container is owned by OrtSessionOptions, but internally it would actually be owned by InferenceSession.

Reply:
Not quite following. Ownership is outside of InferenceSession isn't it? 

Providing the container via OrtSessionOptions is consistent with how we handle shared initializers, so from a user perspective isn't that less confusing? 

And regardless of how it's passed in, as this is a) optional, and b) isn't needed at construction time I think it should be set via a separate method and not a ctor arg to minimize the impact on the InferenceSession API.

Reply:
Sure, we could set it as using an internal method without impacting the internal InferenceSession API  , but how would the flow look in the public API ? The user adds it to OrtSessionOptions and a method is called on the internal InferenceSession instance to pass on a pointer to the container to InferenceSession to use while it is initializing ?

Is it okay for that the container that still lives in OrtSessionOptions given that we take in const OrtSessionOptions instance in the public APIs and to have a container (that lives in OrtSessionOptions) to get filled-in/mutated ?

Reply:
No. Mutating a container in OrtSessionOptions that is passed as const-ptr is definitely not what we want to do. Since the public session creation API includes both model loading and session initialization, it's ok to keep the public API as-is. Internally, call the Set-API on the InferenceSession instance inside 'OrtApis::CreateSession'.

Reply:
The public API is as is, the internal InferenceSession class now has the recommended change